### PR TITLE
refactor: Principal overhaul

### PR DESCRIPTION
**Pull request by lwshang**
This refactor introduced breaking changes, so bumped up version to `0.4.0-beta.0`.

## Breaking changes
* Remove `PrincipalInner`
  * `Principal` directly holds `len` and `bytes` fields
* `PrincipalError` enum has different set of variants reflecting changes in `from_text` logic.
* `from_text` accepts input containing uppercase letters which results in Err before.
* `from_text` verifies CRC32 check sequence

## Implementation details
* Mark `Principal::from_slice` as deprecated
  * It panics if input longer than 29 bytes
  * recommend to use `try_from_slice` 
* Move all tests to root/tests folder so that we only test public visible behavior
* Use `data-encoding` instead of `base32` 
  * `base32` wrongly accepts invalid base32 string which makes error handling in `from_text` not clear
  * `data-encoding` is better maintained and has more dependents and downloads
* Use `serde_test`
  * `serde_json` and `serde_cbor` are kept in `dev-dependencies` because of doc-tests
* Use `impl::impl!` macro to verify the necessary traits are implemented

## Discussion about data layout
```rust
struct Principal(PrincipleInner);
#[repr(packed)]
struct PrincipleInner {
  len: u8,
  bytes: [u8; 29],
}
```
After removing the layer of `PrincipalInner`, it becomes:
```rust
#[repr(packed)]
struct Principal {
  len: u8,
  bytes: [u8; 29],
}
```
I'm not sure whether we should keep the `repr(packed)` here. The [Rust Nomicon ](https://doc.rust-lang.org/nomicon/other-reprs.html#reprpacked) says:
> repr(packed) forces Rust to strip any padding, and only align the type to a byte. This may improve the memory footprint, but will likely have other negative side-effects.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8b19c6d46a7600f575cc936c9dd746ca
With or without `repr(packed)`, the type always has size of 30.

### CODE REVIEW

File: src/principal.rs [hunk 1]
```diff
@@ -113,131 +84,155 @@ impl TryFrom<u8> for PrincipalClass {
 /// );
 /// ```
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct Principal(PrincipalInner);
+#[repr(packed)]
```
Comment (ericswanson-dfinity) [inline src/principal.rs:87]: Relating to what you brought up in the PR description, I think unless we can accompany this with some plausible description of "extreme requirements," we should remove `#[repr(packed)]` here.

This is from the [Rust Nomicon](https://doc.rust-lang.org/nomicon/other-reprs.html#reprpacked):
> repr(packed) is not to be used lightly. Unless you have extreme requirements, this should not be used.

**Comment by lwshang** [comment]
I created a [repo](https://github.com/lwshang/principal-check) to illustrate that with or without `#[repr(packed)]`, `Principal` always has size of 30 on wasm32 platform.
So it should be safe to remove the annotation.

I also made a [draft PR](https://github.com/dfinity/candid/pull/357) to `candid` which makes sure that `candid` won't be broken.

File: src/principal.rs (hunk 1 above)
Comment (lwshang) [reply src/principal.rs]: Removed after verification.

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+
+### Changed
+
+- `PrincipalError` enum has different set of variants reflecting changes in `from_text` logic.
+- `from_text` accepts input containing uppercase letters which results in Err before.
+- `from_text` verifies CRC32 check sequence
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:14]: Should these mention the context (Principal)?

Comment (ericswanson-dfinity) [inline CHANGELOG.md:14]: Should these mention the context (Principal)?

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
+
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:15]:
```suggestion
- `Principal::from_slice` is deprecated in favor of `Principal::try_from_slice`
```

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**